_posts/2023-05-10-pact-and-quarkus-3.adoc

```diff
@@ -0,0 +1,105 @@
+
+== Contract test options
+
+There are a few different contract-testing frameworks out there,
+including Pact, Microcks, Spring Cloud Contract.
+Some teams also build up their own OpenAPI-based toolchains, such as Schemathesis for functional tests, and Prism for the mocking.
+Arguably the most popular contract testing solution is Pact, so it's where the Quarkiverse support for contract testing has started.
+
+- Pact is polyglot, with bindings for almost all popular languages.
+- It's an integrated solution which provides both mocks for consumers and functional tests for providers.
+- It's standalone, and can be run without standing up any extra services, although a Pact Broker with some nice value-adds is available.
+- Although Pact started as a REST-only solution, it is now https://docs.pact.io/blog/2022/11/11/pact-plugin-framework-launch[pluggable], which allows it to support a range of protocols and transports
+
+The Pact team have a https://pactflow.io/blog/contract-testing-using-json-schemas-and-open-api-part-1/[good overview] of the advantages and disadvantages of schema-based testing
+(such as validation based on an OpenAPI spec) and contract testing.
+
+== What's new with Pact and Quarkus
+
+Using Pact with Quarkus isn't new; Quarkus contributors made https://github.com/quarkusio/quarkus/commit/70902a005842a083814aea13567b8bc82574a021[several] https://github.com/quarkusio/quarkus/commit/910227622e0e4c51c111872962cfd5364674673a#diff-40752b74da493b2aa06facf4b1ebd7a81736f6318dcfe76ddae90deafa57fad7[classloading adjustments] in Quarkus core to support Pact testing in Quarkus 2.0,
+but this support was limited. In particular, Pact tests couldn't run in continuous testing mode.
+
+Quarkus 3.0 moves Pact support from Quarkus core to its own Quarkiverse extension, where it can be deeper.
+Quarkus core also includes classloading changes in the Kotlin extension and some classloading fixes in continuous testing itself.
+These mean that, with the Pact Quarkiverse http://quarkus.io/extensions/io.quarkiverse.pact/quarkus-pact-provider[provider]
+and http://quarkus.io/extensions/io.quarkiverse.pact/quarkus-pact-consumer[consumer extensions] Pact tests work properly with `quarkus test` and `quarkus dev`.
+
+To install the consumer extensions, run
+
+```
+quarkus ext add io.quarkiverse.pact:quarkus-pact-consumer
+```
+
+The provider extension can be installed with
+```
+quarkus ext add io.quarkiverse.pact:quarkus-pact-provider
+```
+
+For a deeper dive into contract testing, check out https://www.youtube.com/watch?v=d9CSY8HuZ9U[Quarkus Insights #117].
+
+
+== Summary
+
+If you're using microservices, you should seriously consider contract testing.
+With the new Pact extension, Quarkus 3 allows contract tests to be developed using the same
+great workflow as other tests.
+
+== More resources
+
+- https://github.com/holly-cummins/pact-quarkus-sweater-demo[Sample application showing the Quarkus Pact extension]
```

Comment:
Is there a quickstart guide or more generally a guide on how to use the extension that can be added here as well?

Reply:
Awkwardly, no. There should be, clearly, but ... time.

Reply:
The main docs is the readme, so I should perhaps copy that across to a Guide and then that solves the missing-guide problem.

Reply:
+1